controllers/dspipeline_controller.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	mf "github.com/manifestival/manifestival"
+	dspipelinesiov1alpha1 "github.com/opendatahub-io/ds-pipelines-controller/api/v1alpha1"
+	"github.com/opendatahub-io/ds-pipelines-controller/controllers/config"
+	routev1 "github.com/openshift/api/route/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const finalizerName = "dspipelines.opendatahub.io/finalizer"
+
+// DSPipelineReconciler reconciles a DSPipelineParams object
+type DSPipelineReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+func (r *DSPipelineReconciler) Apply(owner mf.Owner, params *DSPipelineParams, template string, fns ...mf.Transformer) error {
+	tmplManifest, err := config.Manifest(r.Client, template, params)
+	if err != nil {
+		return fmt.Errorf("error loading template yaml: %w", err)
+	}
+	tmplManifest, err = tmplManifest.Transform(
+		mf.InjectOwner(owner),
+	)
+
+	tmplManifest, err = tmplManifest.Transform(fns...)
+	if err != nil {
+		return err
+	}
+
+	if err = tmplManifest.Apply(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *DSPipelineReconciler) ApplyWithoutOwner(params *DSPipelineParams, template string, fns ...mf.Transformer) error {
```

Comment:
In what situation would we want to apply without owner?

Reply:
cluster scope resources cannot be owned by namespace scoped resources (even though you can add them, they don't get garbage collected)